core/src/main/java/org/kiwix/kiwixmobile/core/main/TableDrawerAdapter.kt

```diff
@@ -0,0 +1,135 @@
+ *
+ */
+
+package org.kiwix.kiwixmobile.core.main
+
+import android.graphics.Typeface
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView.Adapter
+import androidx.recyclerview.widget.RecyclerView.ViewHolder
+import butterknife.BindView
+import butterknife.ButterKnife
+import org.kiwix.kiwixmobile.core.R
+import org.kiwix.kiwixmobile.core.R2
+
+class TableDrawerAdapter constructor(private val listener: TableClickListener) :
+  Adapter<ViewHolder>() {
+
+  private var title: String = ""
+  private var selectedPosition: Int = 0
+  private val sections: MutableList<DocumentSection> = mutableListOf()
+
+  fun setSections(sections: List<DocumentSection>) {
+    selectedPosition = 0
+    this.sections.clear()
+    this.sections.addAll(sections)
+  }
+
+  fun setTitle(title: String) {
+    this.title = title
+  }
+
+  override fun getItemViewType(position: Int): Int {
+    return if (position == 0) 0
+    else 1
+  }
+
+  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TableDrawerViewHolder {
+    val res = R.layout.section_list
+    val context = parent.context
+    val view = LayoutInflater.from(context).inflate(res, parent, false)
+    if (viewType == 0) return HeaderTableDrawerViewHolder(view)
+    return SectionTableDrawerViewHolder(view)
+  }
+
+  override fun getItemCount(): Int = sections.size + 1
+
+  override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+    val vh = holder as TableDrawerAdapter.TableDrawerViewHolder
+    val context = holder.itemView.context
+    vh.itemView.isActivated = holder.adapterPosition == selectedPosition
+    when (position) {
+      0 -> {
+        vh.textViewTitle.typeface = Typeface.DEFAULT_BOLD
+
+        vh.itemView.setOnClickListener { v: View ->
+          updateSelection(holder.getAdapterPosition())
+          listener.onHeaderClick(v)
+        }
+
+        if (title.isNotEmpty()) vh.textViewTitle.text = title
+        else {
+          var empty = context.getString(R.string.no_section_info)
+          if (context is WebViewProvider) {
+            empty =
+              context.getCurrentWebView()?.title ?: context.getString(R.string.no_section_info)
+          }
+          vh.textViewTitle.text = empty
+        }
+      }
+      else -> {
+        val sectionPosition = position - 1
+        val documentSection: DocumentSection = sections[sectionPosition]
+        vh.textViewTitle.text = documentSection.title
+        val density = context.resources.displayMetrics.density
+        val padding = ((documentSection.level - 1) * PADDING_ * density).toInt()
+        vh.textViewTitle.setPadding(padding, 0, 0, 0)
+        vh.textViewTitle.text = sections[sectionPosition].title
+        vh.itemView.setOnClickListener { v: View? ->
+          updateSelection(vh.adapterPosition)
+          listener.onSectionClick(v!!, sectionPosition)
```

Comment:
Please remove double-bangs(`!!`), double-bangs claim that this var can never be null, but if they are null, then when invoked, it throws an npe and crashes the app, therefore to play on the safer side, we use elvis operator(`?`). You may read about [null-safety in Kotlin](https://kotlinlang.org/docs/reference/null-safety.html)